### ISSUE TITLE
Question about paper

### ISSUE THREAD

**Issue by super233** [BinhuiXie/SPCL#1]
Hi, thanks for your awesome work on contrastive learning in UDA.

I have some questions about the T-SNE visualization in Figure 4 as follows:

1. It seems that the T-SNE visualization results are about pixels in the shown image but not the whole target domain dataset, is my understanding correct? 
2. How can I apply the T-SNE visualization for the whole target domain dataset? Is it feasible to use feature prototype of each category in each target image?
3. Can you provide the code about T-SNE visualization, it's very grateful. :-)

**Comment by BinhuiXie** [comment]
Hello, @super233. Thank you for attention.

1. The T-SNE visualization results in Fig. 4 show the pixels in a single target image.
2. Using feature prototype of each class in each target image is practicable to apply the T-SNE on the whole target domain.
3. We greatly recommend the package of ``sklearn.manifold.TSNE``. And we employ the following setting:
```
TSNE(n_components=2, init='pca', random_state=0)
```

You might also be interested in our recently released work at https://arxiv.org/abs/2105.05013 and please check this rep https://github.com/BIT-DA/SDCA.

**Comment by super233** [comment]
Ok, thanks very much.

**Comment by super233** [comment]
Hi, sorry to disturb you again.

This time the question is about the implementation details, for feature extraction, you modified DeepLab V2 as follows:
> Following [32, 46, 68], sampling rates are fixed as {6, 12, 18, 24} and we modify the stride and dilation rate of the last layers to produce denser feature maps with larger field-of-views.

Counld you please provide a more detailed description of your network? Are the features extracted from the backbone, or the middle layer of ASPP? Have you designed a projector for feature extraction?

It will be very appreciated if the code snippets are available. :-)

**Comment by super233** [comment]
One more question.

Compared with input images, the features extracted always have smaller resolutions, making a point in features corresponds to a region in the input image, but not a pixel. 

Does it have a big impact on the computing and updating of category prototypes? If yes, how do you resolve it?

I'm a green hand in contrastive learning and unsupervised domain adaptation. I'm sorry to heavily disturb you.

**Comment by BinhuiXie** [comment]
Hi, @super233. 

- This code is heavily borrowed from [FADA](https://github.com/JDAI-CV/FADA). Also, you can ref to https://github.com/BIT-DA/SDCA.
- The feature representations for contrastive learning is exactly extracted from the backbone, i.e., shape [B, 2048, h, w]. Actually, an extract projector might be more efficient since the dimension of feature can be reduced via `1x1 Conv` as standard contrastive learning.
- Yes, a pixel (i, j) in the feature space relates to a region in an input. We down-sample the ground-truth labels to obtain the semantic label in the feature space.

**Comment by super233** [comment]
Thanks a lot. :-)
